virtual_rainforest/main.py

```diff
@@ -0,0 +1,128 @@
+"""Defines the function used to run a full simulation of the model.
+
+As well as setting up the function to run the overall virtual rainforest simulation,
+this script also defines the command line entry points for the model.
+"""
+
+from typing import Any, Type, Union
+
+from virtual_rainforest.core.config import validate_config
+from virtual_rainforest.core.logger import LOGGER, log_and_raise
+from virtual_rainforest.core.model import MODEL_REGISTRY, BaseModel, InitialisationError
+
+
+def select_models(model_list: list[str]) -> list[Type[BaseModel]]:
+    """Select the models to be run for a specific virtual rainforest simulation.
+
+    This function looks for models from a list of models, if these models can all be
+    found in the registry then they are returned. Otherwise an error is logged, which
+    should be handled appropriately downstream.
+
+    Args:
+        model_list: A list of models to select
+
+    Raises:
+        InitialisationError: If one or more models cannot be found in the registry
+    """
+
+    # Remove "core" from model list as it is not a model
+    model_list_ = set(model_list) - {"core"}
+
+    LOGGER.info(
+        "Attempting to configure the following models: %s" % sorted(model_list_)
+    )
+
+    # Make list of missing models, and return an error if necessary
+    miss_model = [model for model in model_list_ if model not in MODEL_REGISTRY.keys()]
+    if len(miss_model):
+        log_and_raise(
+            f"The following models cannot be configured as they are not found in the "
+            f"registry: {miss_model}",
+            InitialisationError,
+        )
+
+    # Then extract each model from the registry
+    modules = [MODEL_REGISTRY[model] for model in model_list_]
+
+    return modules
+
+
+def configure_models(
+    config: dict[str, Any], model_list: list[Type[BaseModel]]
+) -> list[BaseModel]:
+    """Configure a set of models for use in a `virtual_rainforest` simulation.
+
+    Args:
+        config: The full virtual rainforest configuration
+        modules: A set of models to be configured
+
+    Raises:
+        InitialisationError: If one or more models cannot be properly configured
+    """
+
+    # Use factory methods to configure the desired models
+    failed_models = []
+    models_cfd = []
+    for model in model_list:
+        try:
+            models_cfd.append(model.from_config(config))
+        except InitialisationError:
+            failed_models.append(model.name)
+
+    # If any models fail to configure inform the user about it
+    if len(failed_models):
```

Comment:
This is really picky and a general question for @dalonsoa and @alexdewar. 

The 'pythonic' approach here is `if failed_models:` because the empty list == False. I think there is a general preference for these 'implicit' tests (https://google.github.io/styleguide/pyguide.html#214-truefalse-evaluations).

I think this is cleaner but I tend to make an exception for `None` - `if var is None:` rather than `if var:` because it is usually in cases where I _explicitly_ want to check that something is not None rather than some other 'falsey' value.

We should aim for a common style here.

Reply:
I personally prefer `if failed_models` over `if len(failed_models)`. I think you're right that consistency is the main thing though.

Reply:
I definitely prefer `if failed_models` - but I'm also trying to justify my liberal use of `if var is None:` elsewhere! 

There is an argument to be made for using explicit testing of truthy/falsey objects when they might take a range of falsey values, but ending up in that situation feels like a code smell. I guess I tend to use `if var is None:` to check for `var` being undefined (default state/optional), which I think of as being different from 'falsey' and so want to make that explicit.

Reply:
In my case, I would normally go for the more verbose `if len(failed_models) == 0` because it is totally explicit and self explanatory. I'm not sure why y suggested the shortened version...

Anyway, consistency, as you and Alex are suggesting, is the important bit: choose one style; stick with it.

Reply:
OK - so to avoid hanging this up too much, lets go with implicit as in the Google style guide, with an exception for `if var is None:` for explicit semantic testing for _undefined_ . I realise that's a split decision. 🍰  and 🍰 👉 😮 .